documentation/writing_tests.md

```diff
@@ -14,6 +19,405 @@
+
+- [Mocha](https://www.npmjs.com/package/mocha), which offers us a well known and stable test framework for both NodeJS and browser tests.
+- [Chai](https://www.npmjs.com/package/chai), a well known assertion library for Node and the browser.
+- [Rollup](https://www.npmjs.com/package/rollup), to bundle JavaScript for different environments, which helps us write a single TypeScript source and trust it will compile correctly for Node and the browser.
+- [Karma](https://www.npmjs.com/package/karma), which allows us to run our tests in multiple browsers.
+
+We've also come up with our own internal tools:
+
+- The [Recorder](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/test-utils/recorder) is a tool that helps us run our live tests against static recordings obtained from a previous successful run, which aims to ensure that our code hasn't changed while benefitting from not having to reach out to real live services.
+
+Now, let's see how we're using the mentioned tools.
+
+### Mocha
+
+Mocha is a well known JavaScript test framework that runs in NodeJS and in the browser. If you're new to Mocha, we recommend you to dive into [Mocha's documentation](https://mochajs.org/).
+
+In this section, we'll explore how the Azure SDK for JS is using mocha, what versions we're using, how we're configuring it, and other recommendations on how to use this library.
+
+#### Mocha in our dependencies
+
+Since we're using [Rush](https://rushjs.io/), we're forced to use the same version of our packages in each one of the projects inside of this repository. If you want to add `mocha` as a dev dependency to a new project inside of this repository, first make sure you are in the root folder of that project, then you can run the following command:
+
+```
+rush add --dev -p mocha
+```
+
+To enhance the `mocha` experience, we're using the following packages:
+
+- [`@types/mocha`](https://www.npmjs.com/package/@types/mocha), the type definitions for mocha.
+- [`mocha-multi`](https://www.npmjs.com/package/mocha-multi), a way to get multiple reporters working with mocha.
+- [`mocha-junit-reporter`](https://www.npmjs.com/package/mocha-junit-reporter), which produces JUnit-style XML test results.
+- [`nyc`](https://www.npmjs.com/package/nyc) is [Istanbul](https://istanbul.js.org/)'s command line interface.
+- [`esm`](https://www.npmjs.com/package/esm), a popular ECMAScript module loader.
+- [`source-map-support`](https://www.npmjs.com/package/source-map-support), which provides [source map](https://github.com/mozilla/source-map) support for stack traces in node via the [V8 stack trace API](https://github.com/v8/v8/wiki/Stack-Trace-API).
+
+A full `rush add` command that includes `mocha` and all of the previous dependencies follows:
+
+```
+rush add --dev -p mocha @types/mocha mocha-multi mocha-junit-reporter nyc esm source-map-support
+```
+
+We'll explore how we use these dependencies up next.
+
+#### Configuring Mocha
```

Comment:
I believe this section is missing a note on how to use Mocha from VSCode, but I don't ever do that, I use the command line. I can investigate, but perhaps somebody already knows.

Reply:
If you're referring to debugging with VS Code, look at the debug config in storage, we have ways to debug the tests with and without rollup.

https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/storage/storage-blob/.vscode/launch.json#L25

We have the debug configs for samples too, but they might not work right now and might require little tweaks since the samples folder is modified structurally.

Maybe, you can keep a `To Do` for this.